c++/src/Writer.cc

```diff
@@ -0,0 +1,659 @@
+    return *this;
+  }
+
+  CompressionKind WriterOptions::getCompression() const {
+    return privateBits->compression;
+  }
+
+  WriterOptions& WriterOptions::setEncodingStrategy(EncodingStrategy strategy) {
+    privateBits->encodingStrategy = strategy;
+    return *this;
+  }
+
+  EncodingStrategy WriterOptions::getEncodingStrategy() const {
+    return privateBits->encodingStrategy;
+  }
+
+  WriterOptions& WriterOptions::setCompressionStrategy(
+    CompressionStrategy strategy) {
+    privateBits->compressionStrategy = strategy;
+    return *this;
+  }
+
+  CompressionStrategy WriterOptions::getCompressionStrategy() const {
+    return privateBits->compressionStrategy;
+  }
+
+  WriterOptions& WriterOptions::setWriterVersion(WriterVersion version) {
+    privateBits->version = version;
+    return *this;
+  }
+
+  WriterVersion WriterOptions::getWriterVersion() const {
+    return privateBits->version;
+  }
+
+  WriterOptions& WriterOptions::setPaddingTolerance(double tolerance) {
+    privateBits->paddingTolerance = tolerance;
+    return *this;
+  }
+
+  double WriterOptions::getPaddingTolerance() const {
+    return privateBits->paddingTolerance;
+  }
+
+  WriterOptions& WriterOptions::setMemoryPool(MemoryPool* memoryPool) {
+    privateBits->memoryPool = memoryPool;
+    return *this;
+  }
+
+  MemoryPool* WriterOptions::getMemoryPool() const {
+    return privateBits->memoryPool;
+  }
+
+  WriterOptions& WriterOptions::setErrorStream(std::ostream& errStream) {
+    privateBits->errorStream = &errStream;
+    return *this;
+  }
+
+  std::ostream* WriterOptions::getErrorStream() const {
+    return privateBits->errorStream;
+  }
+
+  WriterOptions& WriterOptions::setEnableStats(bool enable) {
+    privateBits->enableStats = enable;
+    return *this;
+  }
+
+  bool WriterOptions::getEnableStats() const {
+    return privateBits->enableStats;
+  }
+
+  WriterOptions& WriterOptions::setEnableStrStatsCmp(bool enable) {
+    privateBits->enableStrStatsCmp = enable;
+    return *this;
+  }
+
+  bool WriterOptions::getEnableStrStatsCmp() const {
+    return privateBits->enableStrStatsCmp;
+  }
+
+  WriterOptions& WriterOptions::setEnableIndex(bool enable) {
+    privateBits->enableIndex = enable;
+    return *this;
+  }
+
+  bool WriterOptions::getEnableIndex() const {
+    return privateBits->enableIndex;
+  }
+
+  const Timezone* WriterOptions::getTimezone() const {
+    return privateBits->timezone;
+  }
+
+  WriterOptions&  WriterOptions::setTimezone(const std::string& zone) {
+    privateBits->timezone = &getTimezoneByName(zone);
+    return *this;
+  }
+
+  Writer::~Writer() {
+    // PASS
+  }
+
+  class WriterImpl : public Writer {
+  private:
+    std::unique_ptr<ColumnWriter> columnWriter;
+    std::unique_ptr<BufferedOutputStream> compressionStream;
+    std::unique_ptr<BufferedOutputStream> bufferedStream;
+    std::unique_ptr<StreamsFactory> streamsFactory;
+    OutputStream* outStream;
+    WriterOptions options;
```

Comment:
We better keep a copy of WriterOptions within Writer class. Image below use case -

{
    WriterOption option;
    // set writer option
    // ...
    return createWriter(.. , option);
}

The writer object will be holding a destructed WriterOptions upon function returns.